### ISSUE TITLE
n^2 bug in BufferedFile.read()

### ISSUE THREAD

**Issue by aaronmlevy** [paramiko/paramiko#1864]
I tried to copy a 44MB file from an SFTP server using paramiko's BufferedFile.read(). The read takes ~5 min on v 2.7.2, but only 2 seconds with an adjustment in https://github.com/aaronmlevy/paramiko/tree/924d4e4d6fbd98a3ba796fea285d0b621adb2df4.

I've seen this issue reported previously but not fixed. Should I make a PR?

**Comment by mxmlnkn** [comment]
> Should I make a PR?

Yes, please! The [patch](https://github.com/APrioriInvestments/paramiko/commit/924d4e4d6fbd98a3ba796fea285d0b621adb2df4#diff-ad12ce4be6dc236facf88848ad0f8368dd966574668b6c4f18a8c053852d90ebR193) looks ok to me. It is the usual problem of appending fixed size bytes to a result byte buffer, which needs frequent reallocations and n^2 memory copies, vs. collecting all chunks in a list and concatenating them in one go after having read everything.
